airflow-core/src/airflow/jobs/triggerer_job_runner.py

```diff
@@ -192,9 +202,22 @@ class TriggerStateChanges(BaseModel):
         finished: list[int] | None = None
 
 
+def get_discriminator_value(v: Any) -> Any:
+    if isinstance(v, dict):
+        return v.get("kind", v.get("type"))
+    return getattr(v, "kind", getattr(v, "type", None))
+
+
 ToTriggerRunner = Annotated[
-    Union[workloads.RunTrigger, messages.CancelTriggers, messages.StartTriggerer],
-    Field(discriminator="kind"),
+    Union[
+        Annotated[workloads.RunTrigger, Tag("RunTrigger")],
+        Annotated[messages.CancelTriggers, Tag("CancelTriggersMessage")],
+        Annotated[messages.StartTriggerer, Tag("StartTriggerer")],
+        Annotated[ConnectionResult, Tag("ConnectionResult")],
+        Annotated[VariableResult, Tag("VariableResult")],
+        Annotated[ErrorResponse, Tag("ErrorResponse")],
+    ],
+    Field(discriminator=Discriminator(get_discriminator_value)),
```

Comment:
The Dag processor uses this pattern:

```
    Union["DagFileParsingResult", GetConnection, GetVariable],
    Field(discriminator="type"),
]
```
Which is slightly simpler -- can we use that here too?

Reply:
This is what I wanted to do originally, but I realized that `RunTrigger`, `CancelTriggers` and `StartTriggerer` do not have 'type' but 'kind', this is why I'm building a custom discriminator that will try `kind` and fallback to `type`

Reply:
I updated the `kind` field to match `type` of other `ConnectionResult` `VariableResult` etc...

This way they all share a common discrimanator `type` => could remove the custom discrimator function.